### PR TITLE
feat(utils): add utility class for looking up top level cristin units…

### PR DESCRIPTION
**Pull request by torbjokv**
… with complete cache for high frequency and low concurrency tasks (only norwegian units) #NP-46341

### CODE REVIEW

**Comment by torbjokv** [comment]
quick comment: if you question yourself, why not just pick the top node directly from the unit json on the parent_units array, that would be nice, but parent_units (in plural) is not available on the units endpoint, only the specific units/<id> endpoint

**Comment by torbjokv** [comment]
sjekk opp om  CRISTIN_BOT_FILTER_BYPASS_HEADER er nødvendig og legg på retry på http-kall

File: publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        }
+        while (unit.parentUnit() != null) {
+            unit = CristinUnitsUtil.allUnits.get(unit.parentUnit().id());
+        }
+        return Optional.of(unit);
+    }
+
+    private CristinUnit[] getApiData(int pageNum) {
+        try {
+            String response = fetchResponseWithRetry(new URI(apiUri + API_ARGUMENTS + pageNum));
```
Comment (brinxmat) [inline publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:106]: var

File: publication-commons/src/test/java/no/unit/nva/publication/utils/CristinUnitsUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    private UriRetriever uriRetriever;
+
+    @BeforeEach
+    public void setUp() {
+        this.uriRetriever = mock(UriRetriever.class);
+
+            when(uriRetriever.getRawContent(any(), any())).thenAnswer(
+                (Answer<Optional<String>>) invocationOnMock -> {
+                    URI uri = invocationOnMock.getArgument(0);
+                    int lastDigit = Character.getNumericValue(uri.toString().charAt(uri.toString().length() - 1));
```
Comment (brinxmat) [inline publication-commons/src/test/java/no/unit/nva/publication/utils/CristinUnitsUtilTest.java:30]: var

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            return attempt(() -> JsonUtils.dtoObjectMapper.readValue(response, CristinUnit[].class)).orElseThrow();
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String fetchResponseWithRetry(URI requestUri) {
+        var retryRegistry = RetryRegistry.of(RetryConfig.custom().maxAttempts(5).intervalFunction(
+            IntervalFunction.ofExponentialRandomBackoff()).build());
+        var retryWithDefaultConfig = retryRegistry.retry("executeRequestAsync");
```
Comment (torarnet) [inline publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:116]: `executeRequestAsync `?? your supplier is `executeRequest `?

Comment (torarnet) [reply publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:116]: https://github.com/BIBSYSDEV/nva-cristin-service/blob/main/cristin-commons/src/test/java/no/unit/nva/cristin/common/client/ApiClientTest.java

https://github.com/BIBSYSDEV/nva-cristin-service/blob/main/cristin-commons/src/main/java/no/unit/nva/client/RetryConfigProvider.java

Comment (torarnet) [reply publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:116]: If the supplier is named `executeRequest `this should also be applied to registry: `retryRegistry.retry("executeRequest")`

**Review by torarnet — COMMENTED** [review]
See comment

File: publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java (hunk 3 above)
Comment (torarnet) [reply publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:116]: Maybe write a test for it

File: publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                    allUnits.put(unit.id(), unit);
+                }
+                pageNum++;
+            }
+        }
+    }
+
+    private static Optional<CristinUnit> findTopLevelUnitOrSelf(String unitId) {
+        var unit = CristinUnitsUtil.allUnits.get(unitId);
+        if (unit == null) {
```
[the remark this answers is not in the document]
Comment (torbjokv) [reply publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:115]: implicit cast

File: publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String fetchResponseWithRetry(URI requestUri) {
+        var retryRegistry = RetryRegistry.of(RetryConfig.custom().maxAttempts(5).intervalFunction(
+            IntervalFunction.ofExponentialRandomBackoff()).build());
+        var retryWithDefaultConfig = retryRegistry.retry("executeRequestAsync");
+        Supplier<String> supplier = () -> executeRequest(requestUri);
```
[the remark this answers is not in the document]
Comment (torbjokv) [reply publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:137]: no :-)

File: publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java (hunk 3 above)
Comment (torbjokv) [reply publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:136]: copy pasta bug. did a unit test and deleted it afterwards as I don't see any point keeping it :-)

File: publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java [hunk 6]
```diff
@@ -13,21 +14,27 @@
...
+import nva.commons.core.Environment;
 import nva.commons.core.paths.UriWrapper;
+import nva.commons.core.useragent.UserAgent;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
- * Utility class for retrieving all cristin units where you want to do repeated lookups and keep a large cache like
- * migrations and other bulk operations.
+ * Utility class for retrieving all norwegian cristin units where you want to do repeated lookups and keep a large
```
Comment (brinxmat) [inline publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:36]: Uppercase N

File: publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java [hunk 7]
```diff
@@ -113,14 +134,33 @@ private CristinUnit[] getApiData(int pageNum) {
...
-        var retryWithDefaultConfig = retryRegistry.retry("executeRequestAsync");
+        var retryWithDefaultConfig = retryRegistry.retry("executeRequest");
         Supplier<String> supplier = () -> executeRequest(requestUri);
 
         return ofSupplier(Retry.decorateSupplier(retryWithDefaultConfig, supplier)).get();
     }
 
     private String executeRequest(URI requestUri) {
-        return of(() -> uriRetriever.getRawContent(requestUri, APPLICATION_JSON).orElseThrow()).get();
+        return of(() -> attempt(() -> httpClient.send(HttpRequest.newBuilder()
```
Comment (brinxmat) [inline publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java:144]: I don't love the HttpRequest being inline

**Review by brinxmat — APPROVED** [review]
Couple of comments, but approved

**Pull request merged**